mask/eolearn/mask/snow_mask.py

```diff
@@ -0,0 +1,261 @@
+        if self.dilation_size:
+            snow_masks = np.array([binary_dilation(mask, disk(self.dilation_size)) for mask in snow_masks])
+        return snow_masks
+
+    def execute(self, eopatch):
+        raise NotImplementedError
+
+
+class SnowMask(BaseSnowMask):
+    """
+    The task calculates the snow mask using the given thresholds.
+
+    THe default values were optimised based on the Sentinel-2 L1C processing level. Values might not be optimal for L2A
+    processing level
+    """
+
+    NDVI_THRESHOLD = 0.1
+
+    def __init__(self, data_feature, band_indices, ndsi_threshold=0.4, brightness_threshold=0.3, **kwargs):
+        """
+        :param data_feature: EOPatch feature represented by a tuple in the form of `(FeatureType, 'feature_name')`
+            containing the bands 2, 3, 7, 11, i.e. (FeatureType.DATA, 'BANDS')
+        :type data_feature: tuple(FeatureType, str)
+        :param band_indices: A list containing the indices at which the required bands can be found in the data_feature.
+            The required bands are B03, B04, B08 and B11 and the indices should be provided in this order. If the
+            'BANDS' array contains all 13 L1C bands, then `band_indices=[2, 3, 7, 11]`. If the 'BANDS' are the 12 bands
+            with L2A values, then `band_indices=[2, 3, 7, 10]`
+        :type band_indices: list(int)
+        :param ndsi_threshold: Minimum value of the NDSI required to classify the pixel as snow
+        :type ndsi_threshold: float
+        :param brightness_threshold: Minimum value of the red band for a pixel to be classified as bright
+        :type brightness_threshold: float
+        """
+        super().__init__(data_feature, band_indices, **kwargs)
+        self.ndsi_threshold = ndsi_threshold
+        self.brightness_threshold = brightness_threshold
+
+    def execute(self, eopatch):
+        bands = eopatch[self.bands_feature][..., self.band_indices]
+        with np.errstate(divide='ignore'):
+            # (B03 - B11) / (B03 + B11)
+            ndsi = (bands[..., 0] - bands[..., 3]) / (bands[..., 0] + bands[..., 3])
+            # (B08 - B04) / (B08 + B04)
+            ndvi = (bands[..., 2] - bands[..., 1]) / (bands[..., 2] + bands[..., 1])
+
+        ndsi_invalid, ndvi_invalid = ~np.isfinite(ndsi), ~np.isfinite(ndvi)
+        ndsi[ndsi_invalid] = self.undefined_value
+        ndvi[ndvi_invalid] = self.undefined_value
+
+        snow_mask = np.where(np.logical_and(np.logical_or(ndsi >= self.ndsi_threshold,
+                                                          np.abs(ndvi - self.NDVI_THRESHOLD) < self.NDVI_THRESHOLD / 2),
+                                            bands[..., 0] >= self.brightness_threshold), 1, 0)
+
+        snow_mask = self._apply_dilation(snow_mask)
+
+        snow_mask[np.logical_or(ndsi_invalid, ndvi_invalid)] = self.undefined_value
+
+        eopatch.add_feature(FeatureType.MASK, self.mask_name, snow_mask[..., np.newaxis].astype(np.uint8))
```

Comment:
3 comments here:
- This could be written in the new notation, either  `eopatch.mask[feature_type][feature_name] = ...` or  `eopatch.mask[snow_mask_feature] = ...`
- As I understand this task is doing single temporal classification. Therefore it could either work on `FeatureType.DATA` or `FeatureType.DATA_TIMELESS`. Up to this point either feature type would work. Here (or even better - already in `__init__`) the task could decide to store results in either `FeatureType.MASK` of `FeatureType.MASK_TIMELESS`.
- Does the mask have to be cast to `np.uint8`? Since it is a binary mask it seems more natural to store it in a boolean type. `AddCloudMaskTask` is also returning mask values in boolean.

I think everything here also applies to the same case in `TheiaSnowMask` task. 
(Even the second bullet should be valid as you can at the beginning add temporal dimension to timeless data.)

Reply:
changed returned type to `bool`. The `IS_DATA` array is `uint8`, so we should unify all binary masks type to be `bool`.

Changed initialisation of a `snow_feature` in `__init__`.

I think the task should work on `DATA` and return `MASK` only, to avoid confusions. If a single time frame is provided, the first dimension of the output will have size 1. This is how the cloud masking works, and I don't think we should change it.

Reply:
At some point we'll have to update `IS_DATA` dtype as well. Probably when we'll be updating `eo-learn-io` with new SH API.

I guess we can leave snow masking limited to `DATA` features for now. Some time in the future we'll have to unify cloud and snow masking tasks. At that point we can also think about such extensions.